### PR TITLE
HOSTEDCP-1327: Remove Old Azure SDKs from Azure Infra

### PR DESCRIPTION
**Pull request by bryan-cox**
**What this PR does / why we need it**:
The previous Azure infra create and destroy code was using Azure SDKs that are now deprecated. This PR replaces that code with Azure SDKs that are supported and updated.

**Which issue(s) this PR fixes**:
Fixes [HOSTEDCP-1327](https://issues.redhat.com/browse/HOSTEDCP-1327)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: cmd/infra/azure/destroy.go [hunk 1]
```diff
@@ -64,29 +63,27 @@ func (o *DestroyInfraOptions) Run(ctx context.Context) error {
...
 	}
-	authorizer, err := auth.ClientCredentialsConfig{
-		TenantID:     creds.TenantID,
-		ClientID:     creds.ClientID,
-		ClientSecret: creds.ClientSecret,
-		AADEndpoint:  azure.PublicCloud.ActiveDirectoryEndpoint,
-		Resource:     azure.PublicCloud.ResourceManagerEndpoint,
-	}.Authorizer()
+
+	_ = os.Setenv("AZURE_TENANT_ID", creds.TenantID)
```
Comment (enxebre) [inline cmd/infra/azure/destroy.go:67]: why does this need to be set in the env now?

File: cmd/infra/azure/create.go [hunk 2]
```diff
@@ -95,254 +118,404 @@ func readCredentials(path string) (*apifixtures.AzureCreds, error) {
...
 
-	result := CreateInfraOutput{
-		Location:   o.Location,
-		InfraID:    o.InfraID,
-		BaseDomain: o.BaseDomain,
+	return creds.SubscriptionID, azureCreds, nil
+}
+
+// createResourceGroup creates the Azure resource group used to group all Azure infrastructure resources
+func createResourceGroup(ctx context.Context, azureCreds azcore.TokenCredential, subscriptionID string, name string, infraID string, location string) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (bryan-cox) [reply cmd/infra/azure/create.go:243]: All of the calls to create resources did have to be rewritten for the new SDKs. 

I put them into functions:
1. to make it easier to quickly read and understand the flow of events for the `Run` function
2. in hopes of making a follow up PR sometime to unit test them (I believe there is a mock way of doing this but I need to look into it further)

File: cmd/infra/azure/destroy.go (hunk 1 above)
Comment (bryan-cox) [reply cmd/infra/azure/destroy.go:67]: From my understanding, this is how we would authenticate with Azure. It has a process of authenticating a user to get valid Azure credentials for creating/destroying resources.

https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity#readme-key-concepts
https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity#DefaultAzureCredential

File: cmd/infra/azure/create.go [hunk 3]
```diff
@@ -95,254 +118,404 @@ func readCredentials(path string) (*apifixtures.AzureCreds, error) {
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to create RHCOS image: %w", err)
+	}
+
+	if o.OutputFile != "" {
+		resultSerialized, err := yaml.Marshal(result)
+		if err != nil {
+			return nil, fmt.Errorf("failed to serialize result: %w", err)
+		}
+		if err := os.WriteFile(o.OutputFile, resultSerialized, 0644); err != nil {
```
[the remark this answers is not in the document]
Comment (jparrill) [reply cmd/infra/azure/create.go:208]: Ty Bryan:

```
OUTPUT_INFRA_FILE is the file where IDs of the infrastructure that has been created will be stored in JSON format. This file can then be used as input to the hypershift create cluster aws command to populate the appropriate fields in the HostedCluster and NodePool resources.
```
- https://redhat-internal.slack.com/archives/D058EFQ0SLA/p1702398366231689

**Pull request merged**